src/objects.c

```diff
@@ -523,7 +523,8 @@ CK_ULONG p11prov_obj_get_key_size(P11PROV_OBJ *obj)
 void p11prov_obj_to_reference(P11PROV_OBJ *obj, void **reference,
                               size_t *reference_sz)
 {
-    *reference = p11prov_obj_ref_no_cache(obj);
+    /* XXX this would leak memory */
+    *reference = obj;  //p11prov_obj_ref_no_cache(obj);
     *reference_sz = sizeof(P11PROV_OBJ);
```

Comment:
Does not look right to me, what will prevent freeing the object while openssl still holds a pointer to it?

Reply:
I think the handles are still referenced in the internal context object list (as you use it for refresh and similar stuff).

Does the openssl unref the reference somehow when it stops using this? I did not find a code that would take this reference and unref it. This keeps the objects in the memory after we free the provider context.

The other option might be somehow forcing the free on cleanup, regardless the amount of handed references.

Reply:
I think the right thing would be to unref when we know openssl is done with the object.

If you are passing params to a function, just clean up the params after the function is done?

Reply:
Here is the example of the leak from the basic test:
```
Direct leak of 112 byte(s) in 1 object(s) allocated from:                                               
    #0 0x7f44dc0ba6af in __interceptor_malloc (/usr/lib64/libasan.so.8.0.0+0xba6af)                     
    #1 0x7f44dbbaf14c in CRYPTO_zalloc (/lib64/libcrypto.so.3+0x1af14c)                                 
    #2 0x7f44dac69050 in p11prov_obj_new /home/jjelen/devel/pkcs11-provider/src/objects.c:250           
    #3 0x7f44dac6e7c0 in p11prov_obj_from_handle /home/jjelen/devel/pkcs11-provider/src/objects.c:800   
    #4 0x7f44dac70a89 in p11prov_obj_find /home/jjelen/devel/pkcs11-provider/src/objects.c:978          
    #5 0x7f44dac9f11b in store_fetch /home/jjelen/devel/pkcs11-provider/src/store.c:135                 
    #6 0x7f44dac9fc1c in p11prov_store_load /home/jjelen/devel/pkcs11-provider/src/store.c:249          
    #7 0x7f44dbc0232b in OSSL_STORE_load (/lib64/libcrypto.so.3+0x20232b)                               
    #8 0x55fcedc074db in load_key_certs_crls_suppress apps/lib/apps.c:972                               
    #9 0x55fcedc087e6 in load_key_certs_crls apps/lib/apps.c:1099                                       
    #10 0x55fcedc087e6 in load_key apps/lib/apps.c:560                                                  
    #11 0x55fcedbe3a33 in init_ctx apps/pkeyutl.c:549                                                   
    #12 0x55fcedbe3a33 in pkeyutl_main apps/pkeyutl.c:305                                               
    #13 0x55fcedbd9500 in do_cmd apps/openssl.c:418                                                     
    #14 0x55fcedbb9711 in main apps/openssl.c:298                                                       
    #15 0x7f44db84a50f in __libc_start_call_main (/lib64/libc.so.6+0x2750f)                             
    #16 0x7f44db84a5c8 in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x275c8)                      
    #17 0x55fcedbb99e4  (/usr/bin/openssl+0x489e4)                                                      
```
I went ahead to check how the references are managed. We are getting this reference through the store API, which allocates the internal object structure and then saves them to the store context. The `p11prov_store_load` then gives away the references, which might or might not be exported to the `EVP_PKEY`s through the `p11prov_common_load()` function, which references the key again.

I think that by storing the key objects in the store context marks the lifetime of the references handed over to the application. Once the store context is closed, the references should no longer be valid (unless it was exported to EVP_PKEY -- in that case, it is indeed kept alive with the reference counter). This is correctly implemented in `p11prov_store_ctx_free()` which decrements the references to the keys, freeing the unused and keeping the exported (which are still references by key management objects). Before this change, this does not work and keys are left in memory infinitely.

Or do I miss something here?

Reply:
Your argument is convincing.
Would you mind changing the name of the function to p11prov_obj_to_store_reference() and remove the commented parts.
Add a comment that say we intentionally do not take a ref because it is already held by the store.
Then I think we are good.
I want to change the name to make it clear that is is meant to be used only with the store and not as a generic function (the rest already use p11prov_obj_ref() anyway.

Reply:
Renamed function added comment, reordered the last commit so the CI tests are last.